### PR TITLE
Some fixes/improvements to the Schema inference code. 

### PR DESCRIPTION
**Pull request by gramster**
In particular, fixes the issue with creating nested schema.

Issue #595

### CODE REVIEW

File: sources/lib/api/gcp/bigquery/_schema.py [hunk 1]
```diff
@@ -136,53 +136,86 @@ def from_dataframe(dataframe, default_type='STRING'):
...
+      raise Exception('Cannot create a schema from record %s' % str(data))
 
   @staticmethod
-  def from_list(data):
+  def from_record(source):
     """
-    Infer a BigQuery table schema from a list. The list must be non-empty and be a list
-    of dictionaries (in which case the first item is used), or a list of lists. In the latter
-    case the type of the elements is used and the field names are simply 'Column1', 'Column2', etc.
+    Infer a BigQuery table schema from a list of fields or a dictionary. The type of the elements
```
Comment (qimingj) [inline sources/lib/api/gcp/bigquery/_schema.py]: nit: Infer a BigQuery table schema from a single record (row) which can be a list of fields or a dictionary...

Just want to emphasize the difference between from_record and from_data

File: sources/lib/api/gcp/bigquery/_schema.py [hunk 2]
```diff
@@ -190,25 +223,48 @@ def from_data(source):
...
+          bq_schema = Schema._from_list_record(source)
+      elif len(source) and isinstance(source[0], list) and \
+          all([isinstance(l, list) and len(l) == len(source[0]) for l in source]):
+        # A list of lists all of the same length; treat first entry as a list record.
+        bq_schema = Schema._from_record(source[0])
       else:
-        return Schema.from_list(source)
+        # A heterogeneous list; treat as a record.
+        bq_schema = Schema._from_list_record(source)
+    elif isinstance(source, dict):
```
Comment (qimingj) [inline sources/lib/api/gcp/bigquery/_schema.py]: Maybe we can force users to call "from_record" and disallow dict in from_data? It seems clearer to me that "from_data" deals with a table and from_record deals with a row.

File: sources/lib/api/gcp/bigquery/_schema.py [hunk 3]
```diff
@@ -190,25 +223,48 @@ def from_data(source):
...
+        bq_schema = source
+      elif all(isinstance(d, dict) for d in source):
+        if all('name' in d and 'type' in d for d in source):
+          # It looks like a bq_schema; use it as-is.
+          bq_schema = source
+        elif all(len(d) == len(source[0]) for d in source):
+          bq_schema = Schema._from_dict_record(source[0])
+        else:
+          bq_schema = Schema._from_list_record(source)
+      elif len(source) and isinstance(source[0], list) and \
```
Comment (qimingj) [inline sources/lib/api/gcp/bigquery/_schema.py]: len(source) is always greater than 0 here so no need for this condition?

File: sources/lib/api/tests/bq_schema_tests.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+
+class TestCases(unittest.TestCase):
+
+  def test_schema_from_dataframe(self):
+    df = self._create_data_frame()
+    result = gcp.bigquery.Schema.from_data(df)
+    self.assertEqual(gcp.bigquery.Schema.from_data(self._create_inferred_schema()), result)
+
+  def test_schema_from_data(self):
```
Comment (qimingj) [inline sources/lib/api/tests/bq_schema_tests.py]: Nice tests!

**Comment by qimingj** [comment]
A few minor observations. Thanks!

File: sources/lib/api/gcp/bigquery/_schema.py (hunk 1 above)
Comment (gramster) [inline sources/lib/api/gcp/bigquery/_schema.py]: Done

File: sources/lib/api/gcp/bigquery/_schema.py (hunk 3 above)
Comment (gramster) [inline sources/lib/api/gcp/bigquery/_schema.py]: Done

File: sources/lib/api/gcp/bigquery/_schema.py (hunk 2 above)
Comment (gramster) [inline sources/lib/api/gcp/bigquery/_schema.py]: Done this in the three places where we fell back to treating source as a record (not a list, list of heterogeneous dicts/lists).

**Comment by gramster** [comment]
PTAL

**Comment by qimingj** [comment]
LGTM!

**Pull request merged**